executor/index_lookup_hash_join.go

```diff
@@ -0,0 +1,449 @@
+	}
+	return
+}
+
+func (iw *indexHashJoinInnerWorker) fetchInnerResults(ctx context.Context, task *lookUpJoinTask) error {
+	lookUpContents, err := iw.constructLookupContent(task)
+	if err != nil {
+		return err
+	}
+	lookUpContents = iw.sortAndDedupLookUpContents(lookUpContents)
+	if iw.finished.Load().(bool) {
+		return nil
+	}
+	return iw.innerWorker.fetchInnerResults(ctx, task, lookUpContents)
+}
+
+func (iw *indexHashJoinInnerWorker) handleHashJoinInnerWorkerPanic(r interface{}) {
+	if r != nil {
+		iw.resultCh <- &indexHashJoinResult{err: errors.Errorf("%v", r)}
+	}
+	iw.wg.Done()
+}
+
+func (iw *indexHashJoinInnerWorker) handleTask(ctx context.Context, task *indexHashJoinTask, joinResult *indexHashJoinResult) error {
+	iw.wg = &sync.WaitGroup{}
+	iw.wg.Add(1)
+	go util.WithRecovery(func() { iw.buildHashTableForOuterResult(ctx, task) }, iw.handleHashJoinInnerWorkerPanic)
+	err := iw.fetchInnerResults(ctx, task.lookUpJoinTask)
+	if err != nil {
+		return err
+	}
+	iw.wg.Wait()
+	if iw.finished.Load().(bool) {
+		return nil
+	}
+	return iw.doJoin(ctx, task, joinResult)
+}
+
+func (iw *indexHashJoinInnerWorker) doJoin(ctx context.Context, task *indexHashJoinTask, joinResult *indexHashJoinResult) error {
+	var ok bool
+	iter := chunk.NewIterator4List(task.innerResult)
+	for row := iter.Begin(); row != iter.End(); row = iter.Next() {
+		ok, joinResult = iw.joinMatchedInnerRow2Chunk(ctx, row, task, joinResult)
+		if !ok {
+			return errors.New("indexHashJoinInnerWorker.handleTask failed")
+		}
+	}
+	for rowIdx, val := range task.outerRowStatus {
+		if val == outerRowMatched {
+			continue
+		}
+		iw.joiner.onMissMatch(val == outerRowHasNull, task.outerResult.GetRow(rowIdx), joinResult.chk)
+		if joinResult.chk.IsFull() {
+			iw.resultCh <- joinResult
+			joinResult, ok = iw.getNewJoinResult(ctx)
+			if !ok {
+				return errors.New("indexHashJoinInnerWorker.handleTask failed")
+			}
+		}
+	}
+	return nil
+}
+
+func (iw *indexHashJoinInnerWorker) joinMatchedInnerRow2Chunk(ctx context.Context, innerRow chunk.Row, task *indexHashJoinTask,
+	joinResult *indexHashJoinResult) (bool, *indexHashJoinResult) {
+	var err error
+	keyBuf := make([]byte, 0, 64)
+	keyBuf, err = codec.HashChunkRow(iw.ctx.GetSessionVars().StmtCtx, keyBuf, innerRow, iw.rowTypes, iw.keyCols)
+	if err != nil {
+		joinResult.err = err
+		return false, joinResult
+	}
+	iw.matchedOuterPtrs = task.lookupMap.Get(keyBuf, iw.matchedOuterPtrs[:0])
+	if len(iw.matchedOuterPtrs) == 0 {
+		return true, joinResult
+	}
+	innerIter := chunk.NewIterator4Slice([]chunk.Row{innerRow})
+	var ok bool
+	for _, ptr := range iw.matchedOuterPtrs {
+		innerIter.Begin()
+		rowIdx := *(*int)(unsafe.Pointer(&ptr[0]))
+		outerRow := task.outerResult.GetRow(rowIdx)
+		matched, isNull, err := iw.joiner.tryToMatch(outerRow, innerIter, joinResult.chk)
+		if err != nil {
+			joinResult.err = err
+			return false, joinResult
+		}
+		if matched {
+			task.outerRowStatus[rowIdx] = outerRowMatched
+		}
+		if isNull {
```

Comment:
Should be `isNull && task.outerRowStatus[rowIdx] != outerRowMatched`?

Reply:
If `isNull` is true, `matched` will always be false.
So `task.outerRowStatus[rowIdx] != outerRowMatched` will always be true here.